working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,379 @@
+not prevent runtime system from reclaiming the referenced object when it becomes
+_weakly reachable_ or, in other words, reachable only through weak references.
+When an object is dimmed _weakly reachable_ all `WeakRef` objects referencing it
+are cleared and the object is eligible for reclamation.
+
+### Finalization
+
+_Finalization_ is a process of invoking a piece of user defined code at some
+point after certain object has been deemed _unreachable_ by a garbage collector.
+_Unreachable_ objects can no longer be interacted with by any programmatic
+means and thus can be safely reclaimed in a manner otherwise unobservable
+to the rest of the program.
+
+Finalization provides a way to observe and react to the event of runtime system
+discovering that some object is unreachable and can be used in a variety of
+ways. Most commonly finalization is used to cleanup some sort of manually
+managed (native) resources associated with a GC managed objects.
+
+For example, finalization can be used to automatically free native resources
+accessed through `dart:ffi` when Dart objects owning these resources become
+unreachable.
+
+```dart
+/// Wrapper for a native resource (accessed through the [_resource] pointer).
+class Wrapper {
+  final ffi.Pointer<ffi.Void> _resource;
+
+  /// Without automatic finalization [Wrapper] user needs to explicitly
+  /// call [destroy] to free the underlying [_resource], when it is no
+  /// longer needed, leading to a possibility of leaks.
+  void destroy() {
+    malloc.free(_resource);
+  }
+}
+```
+
+### Prior art in Dart VM Embedding API
+
+Dart VM exposes an [Embedding API][dart_api.h] which allows the native
+programmer to associate finalization action(s) with a particular object by
+creating a [_weak persistent handle_][weak handle] or
+a [_finalizable handle_][finalizable handle]: after the garbage collector has
+reclaimed the object referenced by such a handle it would clear the handle and
+invoke a callback function associated with it.
+
+There are two important things to highlight about this API:
+
+- callback is invoked post reclamation and can't resurrect the referenced
+object;
+- callback is not allowed to invoke any Dart code or call most of Embedding API
+methods;
+
+### Design constraints and requirements
+
+Here are the main requirements to finalization API:
+
+- finalization callbacks should not interrupt sequential execution of the
+program in a way which is observable by pure Dart code;
+- there should be a way to minimize the possibility of _premature finalization_,
+that is finalization of the wrapper object while native resources owned by it
+are still in use;
+- when using finalization API for managing native resources developer should be
+able to:
+  - rely on them when writing fully synchronous code;
+  - rely on finalization callbacks to be invoked when isolate is shutting down;
+
+The first requirements is self-explanatory: we would like to avoid departures
```

Comment:
Other requirement is a subset of the API needs to be efficient to implement in JavaScript.